Kernel/Graphics/VirtIOGPU/GPU3DDevice.cpp

```diff
@@ -0,0 +1,376 @@
+{
+    VERIFY(used + 4 <= max);
+    data[used + 0] = encode_command(3, 0, VirGLCommand::SET_VERTEX_BUFFERS);
+    data[used + 1] = stride;
+    data[used + 2] = offset;
+    data[used + 3] = resource.value();
+    used += 4;
+}
+
+[[maybe_unused]] static void encode_gl_viewport(u32* data, size_t& used, size_t max)
+{
+    constexpr float width = 1024;
+    constexpr float height = 768;
+    VERIFY(used + 8 <= max);
+    data[used + 0] = encode_command(7, 0, VirGLCommand::SET_VIEWPORT_STATE);
+    data[used + 1] = 0;
+    data[used + 2] = bit_cast<u32>(width / 2);  // scale_x
+    data[used + 3] = bit_cast<u32>(height / 2); // scale_y
+    data[used + 4] = bit_cast<u32>(0.5f);       // scale_z
+    data[used + 5] = bit_cast<u32>(width / 2);  // translate_x
+    data[used + 6] = bit_cast<u32>(height / 2); // translate_y
+    data[used + 7] = bit_cast<u32>(0.5f);       // translate_z
+    used += 8;
+}
+
+[[maybe_unused]] static void encode_transfer3d_flat(u32* data, size_t& used, size_t max, ResourceID resource, size_t length)
+{
+    constexpr size_t cmd_len = 13;
+    VERIFY(used + cmd_len + 1 <= max);
+    data[used + 0] = encode_command(cmd_len, 0, VirGLCommand::TRANSFER3D);
+    data[used + 1] = resource.value(); // res_handle
+    data[used + 2] = 0;                // level
+    data[used + 3] = 242;              // usage
+    data[used + 4] = 0;                // stride
+    data[used + 5] = 0;                // layer_stride
+    data[used + 6] = 0;                // x
+    data[used + 7] = 0;                // y
+    data[used + 8] = 0;                // z
+    data[used + 9] = length;           // width
+    data[used + 10] = 1;               // height
+    data[used + 11] = 1;               // depth
+    data[used + 12] = 0;               // data_offset
+    data[used + 13] = 1;               // direction
+    used += cmd_len + 1;
+}
+
+[[maybe_unused]] static void encode_transfer3d_2d_res(u32* data, size_t& used, size_t max, ResourceID resource, Protocol::Rect dirty_rect)
+{
+    constexpr size_t cmd_len = 13;
+    VERIFY(used + cmd_len + 1 <= max);
+    data[used + 0] = encode_command(cmd_len, 0, VirGLCommand::TRANSFER3D);
+    data[used + 1] = resource.value();   // res_handle
+    data[used + 2] = 0;                  // level
+    data[used + 3] = 242;                // usage
+    data[used + 4] = 0;                  // stride
+    data[used + 5] = 0;                  // layer_stride
+    data[used + 6] = dirty_rect.x;       // x
+    data[used + 7] = dirty_rect.y;       // y
+    data[used + 8] = 0;                  // z
+    data[used + 9] = dirty_rect.width;   // width
+    data[used + 10] = dirty_rect.height; // height
+    data[used + 11] = 1;                 // depth
+    data[used + 12] = 0;                 // data_offset
+    data[used + 13] = 1;                 // direction
+    used += cmd_len + 1;
+}
+
+[[maybe_unused]] static void encode_end_transfers_3d(u32* data, size_t& used, size_t max)
```

Comment:
Why do we need the `[[maybe_unused]]` tag here?

Reply:
Ah, this was the kernel space code for encoding some command buffers. I'm keeping this in here for now for the above stated reasons, the `[[maybe_unused]]` is that there are currently no users of the code.

Reply:
It appears that we probably shouldn't need this code in the kernel, so I'll remove it. Worst case scenario, I'll just have to copy it across from userspace again.